esp/platform/txsummary.cpp

```diff
@@ -65,100 +398,490 @@ bool CTxSummary::contains(const char* key) const
+            DBGLOG("%s", summary.str());
+        }
     }
 }
 
 bool CTxSummary::__contains(const char* key) const
 {
-    return find_if(m_entries.begin(), m_entries.end(), MATCH_KEY) != m_entries.end();
+    // Because now a key can be a path, we can't just
+    // compare the key to the fullname of all entries.
+    // We need to ensure that adding the new key won't
+    // duplicate or disrupt any part of an already-added
+    // path.
+    //
+    // If either the complete key or the entry fullname
+    // is a prefix of the other on the boundary of a dot
+    // delimiter, then we already have an entry named 'key'
+    // For example:
+    //
+    // entries: foo.bar = 5, foo.baz = 6
+    // key  : foo = 7
+    //
+    // We say it contains foo because foo is the parent entry
+    // holding the bar and baz values. We can't append a
+    // duplicate entry foo with value 7.
+
+    for(const auto& pair : m_entries2)
+    {
+        const char* k = key;
+        const char* e = pair->fullname.get();
+        if( nullptr == k || nullptr == e)
+            return false;
+
+        bool done = false;
+        while(!done)
+        {
+            if('\0' == *k)
+            {
+                done = true;
+                if('\0' == *e || '.' == *e)
+                    return true;
+            }
+            else if('\0' == *e)
+            {
+                done = true;
+                if('\0' == *k || '.' == *k)
+                    return true;
+            }
+            else if(toupper(*k) != toupper(*e))
+            {
+                done = true;
+            }
+            else
+            {
+                k++;
+                e++;
+            }
+        }
+    }
+    return false;
+}
+
+CTxSummary::TxEntryBase* CTxSummary::queryEntry(const char* key)
+{
+    if(!validateKey(key))
+        return nullptr;
+
+    for(auto entry : m_entries2)
+    {
+        if(strieq(key, entry->fullname.get()))
+            return entry.get();
+    }
+
+    return nullptr;
+}
+
+
+void CTxSummaryProfileBase::addMap(const char* name, TxSummaryMapVal mapval)
+{
+    mapping.insert({name, mapval});
+}
+
+// Return true if a mapping was found. Set effectiveName parameter to
+// the effective name- the name to use in serialzed output.
+// effectiveName is set to the new name in the first matching mapping
+// found, otherwise the passed in
+bool CTxSummaryProfileBase::getEffectiveName(StringBuffer& effectiveName, const char* name, const unsigned int outputGroup, const unsigned int outputStyle)
+{
+    auto nameMatches = mapping.equal_range(name);
+    for(auto it = nameMatches.first; it != nameMatches.second; it++)
+    {
+        unsigned int mappingGroup = it->second.group;
+        unsigned int mappingStyle = it->second.style;
+        if((mappingGroup & outputGroup) && (mappingStyle & outputStyle))
+        {
+            effectiveName.set(it->second.newName);
+            return true;
+        }
+    }
```

Comment:
Return a matching iterator or end(). Act on the iterator here and in queryEffectiveName.